### ISSUE TITLE
Basic monitoring

### ISSUE THREAD

**Issue by celskeggs** [sipb/homeworld#180]
As part of #172, and as a step towards #16, we need to set up some infrastructure that can monitor different parts of a cluster, either during setup or production.

At some point, this will plug into Prometheus, probably, but we need a broader system, so that we have something that works before we can launch prometheus.

This will monitor elements that include, but are not limited to:

  * That each node is online.
  * ~That each systemd service is running on each node.~ deferred, since this isn't necessary to replace the existing verification scripts.
  * That etcd is healthy.
  * That kubernetes is healthy.
  * That containers can be pulled from the registry.
  * That the core cluster services (flannel, dns-addon) are actively running.
  * That the flannel overlay network is functioning.
  * That DNS is actively accessible (via the dns-addon).
  * That SSH access is possible via cluster auth.

**Comment by celskeggs** [comment]
My initial thought for our monitoring system looks like this:

A monitoring daemon monitors a set of services through various mechanisms and reports everything in a uniform format. It shouldn't try to pass along detailed information about *what* is broken, or try particularly hard to diagnose the problem, though a small amount of information on exactly how the system *knows* something is broken should be available, as a place to start for debugging.

I'm considering structuring the system as a set of "flags" -- individual tracked markers that can be in one of a set of states: healthy, degraded, and unhealthy (green flag/yellow flag/red flag). The state of the cluster as a whole is based on the worst flag raised.

Roughly speaking, the cluster being unhealthy means "there is a problem *now* and someone should take care of it ASAP", whereas degraded means "there are warning signs that a problem might materialize, but it can wait". It might be the case that yellow flags should be automatically upgraded to red flags after some amount of time, but I'm not sure.

Every monitor should be *continuous*, unlike some of the current verification steps -- if something breaks, a flag should be raised in less than thirty seconds -- and preferably in only a few seconds. If a flag is oscillating between more than one state, it should stay at the worse state. This means that if a flag returns to healthy/degraded after being degraded/unhealthy, it should remain in the worse state but be marked as "recovering" until enough time passes that it can be counted as the better state.

All status information should be made available by the monitoring daemon over a simple HTTP connection that produces JSON. Higher level interfaces (like web interfaces) should be in separate processes, for reliability.

(At some point, the monitoring interface should be upgraded to be HTTPS, so that an attacker can't spoof a healthy state for the cluster. But, to start, HTTP will suffice.)

For reference, I haven't ever deployed monitoring for a production system, so I'd be very interested in suggestions about these thoughts.

**Comment by celskeggs** [comment]
ikdc's observation from zephyr is accurate: it makes more sense to solve the problems of having pre-cluster prometheus than the problems of a more hand-rolled monitoring system.
